src/xhr.js

```diff
@@ -1,31 +1,80 @@
 import { fetch } from './fetch.js';
 import { Evented } from './evented.js';
 
+// Extends from Evented to fire events
 export class XMLHttpRequest extends Evented {
-  open (method, url) {
+  constructor() {
+    super();
+    this.headers = {};
+    this.readyState = 1;
+    this.withCredentials = true;
+  }
+
+  // Stores method and url
+  open(method, url) {
     this.method = method;
     this.url = url;
   }
-  send () {
-    // print('XHR', this.method, this.url);
-    fetch(this.url, { method: this.method })
+
+  /**
+   * Resets status and values
+   * Needed by some implementations
+   */
+  abort() {
+    this.responseText = '';
+    this.response = null;
+    this.readyState = 0;
+    this.status = 0;
+  }
+
+  // Append header to headers object
+  setRequestHeader(...args) {
+    this.headers[args[0]] = args[1];
+  }
+
+  // Returns all headers
+  getAllResponseHeaders() {
+    return this.responseHeaders;
+  }
+
+  /**
+   * Uses fetch to fire request
+   * Fires both load and readystatechange so registered components know data was loaded
+   */
+  send(body) {
+    fetch(this.url, { method: this.method, body })
       .then(res => {
+        this.readyState = 4;
         this.status = res.status;
-        // print('XHR', this.status, this.url);
         this.statusText = res.statusText;
         this.redirected = res.redirected;
-        this.headers = res.headers;
+        this.responseHeaders = res.headers;
+
+        let responseBody;
         if (this.responseType === 'arraybuffer') {
-          this.bodyField = 'response';
-          return res.arrayBuffer();
+          responseBody = res.arrayBuffer();
         } else {
-          this.bodyField = 'responseText';
-          return res.text();
+          responseBody = res.text();
         }
+
+        return responseBody;
       })
-      .then(body => {
-        this[this.bodyField] = body;
-        this.emit('load', {});
-      }, err => this.emit('error', err));
+      .then(
+        body => {
+          this.response = body;
+          this.responseText = body;
```

Comment:
The new code assumes text body and removes the ability to have binary bodies.

Reply:
So after doing some research, I found xhr being implemented in different ways, meaning, some implementations would read from 'response' and some from 'responseText' in both cases expecting text. So what I did was to make it so that no matter what you're reading you'll get the data, whether it's binary or not (since it's defined in the previous step lines 54 - 58). We can talk about it later in the meeting at 1

Reply:
So my suggestion to merge both our codes on this is:

Keep this section as is, you don't need to change this:
```js
if (this.responseType === 'arraybuffer') {	        if (this.responseType === 'arraybuffer') {
          this.bodyField = 'response';	          responseBody = res.arrayBuffer();
          return res.arrayBuffer();	
        } else {	        } else {
          this.bodyField = 'responseText';	          responseBody = res.text();
          return res.text();	
        }	
```

And then later, simply add `this.response = body;` and your new event like this:

```js
        this.response = body;
        this[this.bodyField] = body;
        this.emit('readystatechange', body);
        this.emit('load', {});
...

Reply:
Note that calling `res.text()` on a body that isn't valid UTF-8 will cause an exception, so we don't want to call it when `responseType` has explicitly been set to `'arraybuffer'`.